File: overlay/home/bzzbzz/a/shift.py
import gpiod
import time

def setup_gpio(chip_name, pins):
    """GPIO lábak inicializálása."""
    chip = gpiod.Chip(chip_name)
    lines = {name: chip.get_line(pin) for name, pin in pins.items()}

    for line in lines.values():
        line.request(consumer="shift_register", type=gpiod.LINE_REQ_DIR_OUT)

    return lines

def shift_out(lines, data):
    """
    Shift registerre ír egy 8-bites adatot.
    :param lines: A GPIO lábak.
    :param data: 8-bites számsor, amit a shift registerre kell írni.
    """
    data_pin = lines['data']
    clock_pin = lines['clock']
    latch_pin = lines['latch']

    # Latch reset (zárva tartás adatírás közben)
    latch_pin.set_value(0)
    time.sleep(0.001)

    # Bitenkénti adatkiírás
    for bit in data:
        data_pin.set_value(int(bit))  # Adat beállítása
        time.sleep(0.001)            # Rövid késleltetés
        clock_pin.set_value(1)       # Órajel impulzus
        time.sleep(0.001)
        clock_pin.set_value(0)
    
    # Latch beállítása (adat érvényesítése)
    latch_pin.set_value(1)
    time.sleep(0.001)
    latch_pin.set_value(0)

def main():
    chip_name = input("Adja meg a GPIO chip nevét (pl. gpiochip0): ")
    data_pin = int(input("Adja meg az adat pin számát: "))
    clock_pin = int(input("Adja meg az órajel pin számát: "))
    latch_pin = int(input("Adja meg a tároló pin számát: "))

    # GPIO lábak inicializálása
    pins = {"data": data_pin, "clock": clock_pin, "latch": latch_pin}
    lines = setup_gpio(chip_name, pins)

    while True:
        user_input = input("Adjon meg egy 8-bites bináris számsort (pl. 10101010, vagy 'q' kilépéshez): ")
        if user_input.lower() == 'q':
            break
        if len(user_input) != 8 or not all(c in '01' for c in user_input):
            print("Hiba: Csak 8-bites bináris számot adjon meg!")
            continue
        
        shift_out(lines, user_input)
        print(f"Az adat kiírásra került a shift registerre: {user_input}")

    print("Program vége. GPIO felszabadítása.")
    for line in lines.values():
        line.release()

if __name__ == "__main__":
    main()

File: br2apps/package/bence/shift.c
#include <stdio.h>
#include <gpiod.h>
#include <unistd.h>
#include <string.h>

void shift_out(struct gpiod_line *data_line, struct gpiod_line *clock_line, struct gpiod_line *latch_line, const char *data) {
    // Latch reset (zárva tartás adatírás közben)
    gpiod_line_set_value(latch_line, 0);
    usleep(1000); // 1ms késleltetés

    // Bitenkénti adatkiírás
    for (int i = 0; i < 8; i++) {
        int bit = data[i] - '0'; // Bináris érték (0 vagy 1)
        gpiod_line_set_value(data_line, bit); // Adat kimenet
        usleep(1000); // Rövid késleltetés
        gpiod_line_set_value(clock_line, 1); // Órajel impulzus
        usleep(1000);
        gpiod_line_set_value(clock_line, 0);
    }

    // Latch beállítása (adat érvényesítése)
    gpiod_line_set_value(latch_line, 1);
    usleep(1000);
    gpiod_line_set_value(latch_line, 0);
}

int main() {
    char chip_name[32];
    int data_pin, clock_pin, latch_pin;
    char user_input[9]; // 8 bites bináris adat + 1 hely a lezáró nullának

    // GPIO chip és pinek bekérése
    printf("Adja meg a GPIO chip nevét (pl. gpiochip0): ");
    scanf("%31s", chip_name);

    printf("Adja meg az adat pin számát: ");
    scanf("%d", &data_pin);

    printf("Adja meg az órajel pin számát: ");
    scanf("%d", &clock_pin);

    printf("Adja meg a tároló pin számát: ");
    scanf("%d", &latch_pin);

    // GPIO chip és vonalak inicializálása
    struct gpiod_chip *chip = gpiod_chip_open_by_name(chip_name);
    if (!chip) {
        perror("Nem sikerült megnyitni a GPIO chipet");
        return 1;
    }

    struct gpiod_line *data_line = gpiod_chip_get_line(chip, data_pin);
    struct gpiod_line *clock_line = gpiod_chip_get_line(chip, clock_pin);
    struct gpiod_line *latch_line = gpiod_chip_get_line(chip, latch_pin);

    if (!data_line || !clock_line || !latch_line) {
        perror("Nem sikerült elérni a GPIO vonalakat");
        gpiod_chip_close(chip);
        return 1;
    }

    if (gpiod_line_request_output(data_line, "shift_register", 0) < 0 ||
        gpiod_line_request_output(clock_line, "shift_register", 0) < 0 ||
        gpiod_line_request_output(latch_line, "shift_register", 0) < 0) {
        perror("Nem sikerült kimeneti módra állítani a vonalakat");
        gpiod_chip_close(chip);
        return 1;
    }

    while (1) {
        printf("Adjon meg egy 8-bites bináris számsort (pl. 10101010, vagy 'q' kilépéshez): ");
        scanf("%8s", user_input); // Max. 8 karakter olvasása

        if (strcmp(user_input, "q") == 0) {
            break;
        }

        // Validáció: csak bináris számokat fogad el
        int valid = 1;
        for (int i = 0; i < 8; i++) {
            if (user_input[i] != '0' && user_input[i] != '1') {
                valid = 0;
                break;
            }
        }

        if (!valid || strlen(user_input) != 8) {
            printf("Hiba: Csak 8-bites bináris számot adjon meg!\n");
            continue;
        }

        shift_out(data_line, clock_line, latch_line, user_input);
        printf("Az adat kiírásra került a shift registerre: %s\n", user_input);
    }

    // GPIO felszabadítása
    gpiod_chip_close(chip);
    printf("Program vége. GPIO felszabadítása megtörtént.\n");

    return 0;
}

